mmdet3d/models/middle_encoders/voxel_set_abstraction.py

```diff
@@ -0,0 +1,327 @@
+        self.voxel_size = voxel_size
+        self.point_cloud_range = point_cloud_range
+        self.voxel_center_as_source = voxel_center_as_source
+
+        gathered_channels = 0
+
+        if rawpoints_sa_config is not None:
+            self.rawpoints_sa_layer = MODELS.build(rawpoints_sa_config)
+            gathered_channels += sum(
+                [x[-1] for x in rawpoints_sa_config.mlp_channels])
+        else:
+            self.rawpoints_sa_layer = None
+
+        if voxel_sa_configs_list is not None:
+            self.voxel_sa_layers = nn.ModuleList()
+            for voxel_sa_config in voxel_sa_configs_list:
+                cur_layer = MODELS.build(voxel_sa_config)
+                self.voxel_sa_layers.append(cur_layer)
+                gathered_channels += sum(
+                    [x[-1] for x in voxel_sa_config.mlp_channels])
+        else:
+            self.voxel_sa_layers = None
+
+        if bev_features_channels is not None:
+            self.bev_cfg = dict(
+                bev_features_channels=bev_features_channels,
+                bev_scale_factor=bev_scale_factor)
+            gathered_channels += bev_features_channels
+        else:
+            self.bev_cfg = None
+
+        self.point_feature_fusion_layer = nn.Sequential(
+            ConvModule(
+                gathered_channels,
+                fused_out_channels,
+                kernel_size=(1, 1),
+                stride=(1, 1),
+                conv_cfg=dict(type='Conv2d'),
+                norm_cfg=norm_cfg,
+                bias=bias))
+
+    def interpolate_from_bev_features(self, keypoints: torch.Tensor,
+                                      bev_features: torch.Tensor,
+                                      batch_size: int,
+                                      bev_scale_factor: int) -> torch.Tensor:
+        """Gather key points features from bev feature map by interpolate.
+
+        Args:
+            keypoints (torch.Tensor): Sampled key points (N1 + N2 + ..., NDim)
+            bev_features (torch.Tensor): Bev feature map from the first
+                stage (B, C, H, W).
+            batch_size (int): Input batch size.
+            bev_scale_factor (int): Bev feature map scale factor.
+
+        Returns:
+            torch.Tensor: Key points features gather from bev feature
+                map (N1 + N2 + ..., C)
+        """
+        x_idxs = (keypoints[..., 0] -
+                  self.point_cloud_range[0]) / self.voxel_size[0]
+        y_idxs = (keypoints[..., 1] -
+                  self.point_cloud_range[1]) / self.voxel_size[1]
+
+        x_idxs = x_idxs / bev_scale_factor
+        y_idxs = y_idxs / bev_scale_factor
+
+        point_bev_features_list = []
+        for k in range(batch_size):
+            cur_x_idxs = x_idxs[k, ...]
+            cur_y_idxs = y_idxs[k, ...]
+            cur_bev_features = bev_features[k].permute(1, 2, 0)  # (H, W, C)
+            point_bev_features = bilinear_interpolate_torch(
```

Comment:
To check whether `F.grid_sample` can be used to achieve the same function and validate the performance?